Orca - webhook example with script (#1625)
port-labs/port-docs: docs/build-your-software-catalog/custom-integration/webhook/examples/orca-security.md

---

> description: Ingest both historical and realtime alerts on vulnerabilites in Orca Security.

Reviewer: Match to the title below based on the other comment

title: Orca Security
---

> # Ingest historical and realtime vulnerability alerts in Orca Security

Reviewer: Match to the suggested title

This guide shows you how to ingest both historical and realtime alerts in Orca Security. To do this, you will use Port's webhook feature (for realtime alerts) and/or a Python script (for historical data).

For either ingestion types, you will create the `orcaSecurityAlert` blueprint:

<details>
<summary><b>Orca Security Alert blueprint (Click to expand)</b></summary>

```json showLineNumbers
{
  "identifier": "orcaSecurityAlert",
  "description": "A representation of an Orca Security Alert",
  "title": "Orca Security Alert",
  "icon": "Alert",
  "schema": {
    "properties": {
      "description": {
        "type": "string",
        "title": "Description",
        "description": "Description of alert"
      },
      "source": {
        "type": "string",
        "title": "Source",
        "description": "The source of the alert"
      },
      "status": {
        "type": "string",
        "title": "Status",

>         "description": "Current status of the vulnerability on assets"

Reviewer 2: can this also be an enum?

Author: Same as above

      },
      "recommendation": {
        "type": "string",
        "title": "Recommendation",
        "description": "Steps to take to resolve the issue"
      },
      "severity": {
        "type": "string",
        "title": "Severity",
        "description": "Severity of vulnerability on assets"

>       },

Author: I'm not entirely certain of all the possible values so...

>       },

Reviewer 2: <img width="321" alt="Screenshot 2024-10-08 at 1 40 42 PM" src="https://github.com/user-attachments/assets/97e06580-bb70-4e52-9bc3-e2b8d606cba0">

These are the severities I see in Orca, which is consistent with all the CVE standards. Please update it to be an enum

>       },

Reviewer 2: <img width="435" alt="Screenshot 2024-10-08 at 1 48 48 PM" src="https://github.com/user-attachments/assets/2cd95096-97f7-42b9-b87b-89b95f71782d">
Also for alerts, these are the severities

>       },

Reviewer 2: <img width="435" alt="Screenshot 2024-10-08 at 1 48 48 PM" src="https://github.com/user-attachments/assets/2cd95096-97f7-42b9-b87b-89b95f71782d">
Also for alerts, these are the severities

>       },

Reviewer 2: @lordsarcastic actually these enum: low,medium,critical,high and informational are supposed to be under risk level instead of severity

      "riskLevel": {
        "type": "string",
        "title": "Risk Level",
        "description": "Level of risk exposed to by vulnerability"
      },
      "category": {
        "type": "string",
        "title": "Category",
        "description": "Category of alert"
      },
      "alertLabels": {
        "type": "array",
        "title": "Alert Labels"
      },
      "createdAt": {
        "type": "string",
        "title": "Created At",
        "description": "When the alert first appeared",
        "format": "date-time"
      },
      "lastUpdated": {
        "type": "string",
        "title": "Last Updated",
        "description": "When the alert was last updated",
        "format": "date-time"
      },
      "lastSeen": {
        "type": "string",
        "title": "Last Seen",
        "description": "When the alert was last seen",
        "format": "date-time"
      },
      "assetName": {
        "type": "string",
        "title": "Asset Name",
        "description": "Name of the asset affected by the alert"
      }
    },
    "required": []
  },

>   "mirrorProperties": {},

Reviewer: Making sure the existing data in Orca doesn't contain the service where the alert found?

Author: There isn't any explicit information on the service the alert is attached to. It is up to the user to know how to generate the service name from the information provided by Orca.

Reviewer: @lordsarcastic in the original mapping I added to the task we had this mapping to service & repo
![CleanShot 2024-10-09 at 08 47 24@2x](https://github.com/user-attachments/assets/92a1fa83-1e60-4815-a749-f813a50a9d7e)

Why can't we use it?

  "calculationProperties": {},
  "aggregationProperties": {},
  "relations": {}
}
```

</details>

## Ingest realtime alerts using webhooks

1. Head over to the [Data sources](https://app.getport.io/settings/data-sources) page

2. Click on the "+ Data source" icon at the top right corner

3. Go to the "_Webhook_" tab and click on "Custom integration".

4. Fill in the details of the form like so:

> <img src='/img/guides/createOrcaSecurityVulnerabilityAlertWebhook.png' border='1px' />

Reviewer 2: Can we also change the icon to show the more interesting props such as status, source etc

**Note:** Copy the webhook URL displayed by Port in the next step, you'll need it to set up webhooks on the Orca Security side.

5. In the blueprint mapping section, use the following mapping configuration:

<details>

> <summary><b> Orca Security Alert webhook blueprint (Click to expand)</b></summary>

Reviewer: "webhook"?
Why not just "Orca Security Alert blueprint"?

Author: This is the blueprint for the webhook configuration

Reviewer: Got it wasn't sure

```json showLineNumbers

[
  {
    "blueprint": "orcaSecurityAlert",
    "filter": "true",
    "entity": {
      "identifier": ".body.state.alert_id",
      "title": ".body.state.alert_id + '-' + .body.type_string",
      "properties": {
        "description": ".body.description",
        "assetName": ".body.asset_name",
        "source": ".body.source",
        "status": ".body.state.status",
        "recommendation": ".body.recommendation",
        "severity": ".body.state.severity",
        "riskLevel": ".body.state.risk_level",
        "category": ".body.category",
        "alertLabels": ".body.alert_labels",
        "createdAt": ".body.state.created_at",
        "lastUpdated": ".body.state.last_updated",
        "lastSeen": ".body.state.last_seen"
      },

>       "relations": {}

Reviewer: Same as above the original mapping I shared extracted the service & repo as relation

8. Click on "Configure" -> "Create" and enter the following information:
  - **Name:** Enter a suitable name
  - **Trigger URL:** Enter your Port webhook URL here

Leave the **API Key** field blank. Click on "Save"

9. Test the alert by going to the **Alerts** page, click on one of the alerts -> **Take Action** -> **Notifications** -> **Send to webhook** -> **\<Your created webhook\>**.

You should see the newly created alert entity in your Port catalog:

<img src='/img/guides/orcaSecurityVulnerabilityAlertsCatalog.png' border='1px' />

## Ingest historical vulnerability alerts using Python script
In this example you are going to use the provided Python script to fetch data from the Orca Security REST API and ingest it to Port.

### Prerequisites

This example utilizes the same [blueprint](#ingest-historical-and-realtime-vulnerability-alers-in-orca-security) definition from the previous section.

In addition, you require the following environment variables:

- `PORT_CLIENT_ID` - Your Port client id
- `PORT_CLIENT_SECRET` - Your Port client secret

> - `OCTA_SECURITY_API_TOKEN` - Get that by following the steps below

Reviewer: OCTA? not ORCA?

  - Go to **Settings** -> **Users & Permissions** -> **API** -> **API Tokens**
  - Click on **Add API Token**
  - Fill in the details required and select **Shiftleft Alert Manager** for the role.
  - Click on save and copy the API token

> - `OCTA_SECURITY_API_URL` - The host URL of your Orca Security dashboard. See the below table for which URL to use depending on your region:

Reviewer: OCTA? not ORCA?

---
description: Ingest vulnerability alerts from Orca Security using a custom webhook integration
title: Orca Security
---

# Ingest vulnerability alerts from Orca Security

> This guide shows you how to ingest both historical and realtime alerts in Orca Security. To do this, you will use Port's webhook feature (for realtime alerts) and/or a Python script (for historical data).

Reviewer 3: ```suggestion
This guide demonstrates how to ingest both historical and realtime alerts from Orca Security into Port.  
To do this, we will use Port's webhook feature (for realtime alerts) and/or a Python script (for historical data).
```

**Note:** Copy the webhook URL displayed by Port in the next step, you'll need it to set up webhooks on the Orca Security side.

5. In the blueprint mapping section, use the following mapping configuration:

<details>
<summary><b> Orca Security Alert webhook blueprint (Click to expand)</b></summary>

```json showLineNumbers

[
  {
    "blueprint": "orcaSecurityAlert",
    "filter": "true",
    "entity": {
      "identifier": ".body.state.alert_id",
      "title": ".body.state.alert_id + '-' + .body.type_string",
      "properties": {
        "description": ".body.description",
        "assetName": ".body.asset_name",
        "source": ".body.source",
        "status": ".body.state.status",
        "recommendation": ".body.recommendation",
        "severity": ".body.state.severity",
        "riskLevel": ".body.state.risk_level | tostring",
        "category": ".body.category | tostring",
        "alertLabels": ".body.alert_labels",
        "createdAt": ".body.state.created_at",
        "lastUpdated": ".body.state.last_updated",
        "lastSeen": ".body.state.last_seen",
        "repository": ".body.asset_name | split(\"/\") | .[1] // .[0]"
      }
    }
  }
]

```

</details>

> 6. Click on "Save"

Reviewer 3: ```suggestion
6. Click on "Save".
```
